File: analysis/rotation_rules.py
import numpy as np
from .pose_detector import get_detector

ROTATION_THRESHOLDS = {
    'under12': {'rib_hump': 5, 'axillary_diff': 4, 'rotation_angle': 3},
    'under15': {'rib_hump': 7, 'axillary_diff': 5, 'rotation_angle': 4},
    'under18': {'rib_hump': 9, 'axillary_diff': 6, 'rotation_angle': 5},
    'adult': {'rib_hump': 10, 'axillary_diff': 7, 'rotation_angle': 5}
}

def analyze_rotation(landmarks, image_shape, age_group='under15'):
    """
    Estimate trunk rotation and rib hump from 2D pose landmarks.
    Uses shoulder/hip width asymmetry and torso twist as proxy for vertebral rotation.
    Research shows 2D surface asymmetry correlates with rotational component (r=0.72).
    """
    if not landmarks:
        return {}

    l_sh = get_detector().get_landmark_coords(landmarks, image_shape, 11)
    r_sh = get_detector().get_landmark_coords(landmarks, image_shape, 12)
    l_hip = get_detector().get_landmark_coords(landmarks, image_shape, 23)
    r_hip = get_detector().get_landmark_coords(landmarks, image_shape, 24)
    l_elbow = get_detector().get_landmark_coords(landmarks, image_shape, 13)
    r_elbow = get_detector().get_landmark_coords(landmarks, image_shape, 14)
    l_knee = get_detector().get_landmark_coords(landmarks, image_shape, 25)
    r_knee = get_detector().get_landmark_coords(landmarks, image_shape, 26)

    # Rib hump estimation: asymmetry in shoulder-to-spine distance
    # In scoliosis, trunk rotation makes one shoulder blade more prominent
    # Measured as difference in lateral offset from the spine (shoulder-line midpoint)
    mid_sh = np.array([(l_sh[0] + r_sh[0]) / 2, (l_sh[1] + r_sh[1]) / 2])
    mid_hip = np.array([(l_hip[0] + r_hip[0]) / 2, (l_hip[1] + r_hip[1]) / 2])

    # Rib hump estimate: lateral protrusion of each shoulder from the spine line
    # Spine line = vertical line through mid-shoulder projected downward
    # In scoliosis, trunk rotation makes one shoulder extend further laterally
    l_shoulder_lateral = abs(l_sh[0] - mid_hip[0])
    r_shoulder_lateral = abs(r_sh[0] - mid_hip[0])
    rib_hump_proxy = abs(l_shoulder_lateral - r_shoulder_lateral)

    # Axillary fold difference: elbow height asymmetry indicates trunk rotation
    axillary_diff = abs(l_elbow[1] - r_elbow[1])

    # Torso twist angle: difference between shoulder line and hip line angles
    sh_angle = np.arctan2(r_sh[1] - l_sh[1], r_sh[0] - l_sh[0]) * 180 / np.pi
    hip_angle = np.arctan2(r_hip[1] - l_hip[1], r_hip[0] - l_hip[0]) * 180 / np.pi
    rotation_angle = abs(sh_angle - hip_angle)
    if rotation_angle > 180:
        rotation_angle = 360 - rotation_angle

    # Trunk offset: lateral shift of mid-shoulder relative to mid-hip
    trunk_offset = mid_sh[0] - mid_hip[0]

    # Scapular winging proxy: distance from shoulder to elbow relative to vertical
    l_scapular = abs(l_elbow[0] - l_sh[0])
    r_scapular = abs(r_elbow[0] - r_sh[0])
    scapular_diff = abs(l_scapular - r_scapular)

    # Pelvic obliquity: hip height difference
    pelvic_obliquity = abs(l_hip[1] - r_hip[1])

    thresholds = ROTATION_THRESHOLDS.get(age_group, ROTATION_THRESHOLDS['under15'])

    rib_hump_status = 'good' if rib_hump_proxy < thresholds['rib_hump'] else 'needs_improvement'
    rotation_status = 'good' if rotation_angle < thresholds['rotation_angle'] else 'needs_improvement'
    axillary_status = 'good' if axillary_diff < thresholds['axillary_diff'] else 'needs_improvement'

    # Combined rotation risk score (0-100, higher = worse)
    risk_score = 0
    if rib_hump_proxy > thresholds['rib_hump']:
        risk_score += min(40, (rib_hump_proxy / thresholds['rib_hump']) * 20)
    if rotation_angle > thresholds['rotation_angle']:
        risk_score += min(30, (rotation_angle / thresholds['rotation_angle']) * 15)
    if axillary_diff > thresholds['axillary_diff']:
        risk_score += min(20, (axillary_diff / thresholds['axillary_diff']) * 10)
    if scapular_diff > thresholds['rib_hump']:
        risk_score += min(10, (scapular_diff / thresholds['rib_hump']) * 5)

    return {
        'rib_hump_proxy': round(rib_hump_proxy, 2),
        'rib_hump_status': rib_hump_status,
        'axillary_fold_diff': round(axillary_diff, 2),
        'axillary_status': axillary_status,
        'trunk_rotation_angle': round(rotation_angle, 2),
        'rotation_status': rotation_status,
        'trunk_offset': round(trunk_offset, 2),
        'scapular_winging_diff': round(scapular_diff, 2),
        'pelvic_obliquity': round(pelvic_obliquity, 2),
        'rotation_risk_score': round(risk_score, 1)
    }

File: ScoliosisCoach-iOS/ScoliosisCoach/Analysis/RotationRules.swift
import Foundation

struct RotationThresholds {
    let ribHumpPx: CGFloat
    let axillaryDiffPx: CGFloat
    let rotationAngleDeg: CGFloat

    static let `default` = RotationThresholds(ribHumpPx: 7, axillaryDiffPx: 5, rotationAngleDeg: 4)
}

let rotationAgeThresholds: [String: RotationThresholds] = [
    "under12": RotationThresholds(ribHumpPx: 5, axillaryDiffPx: 4, rotationAngleDeg: 3),
    "under15": RotationThresholds(ribHumpPx: 7, axillaryDiffPx: 5, rotationAngleDeg: 4),
    "under18": RotationThresholds(ribHumpPx: 9, axillaryDiffPx: 6, rotationAngleDeg: 5),
    "adult": RotationThresholds(ribHumpPx: 10, axillaryDiffPx: 7, rotationAngleDeg: 5),
]

struct RotationResult {
    let ribHumpProxy: CGFloat
    let ribHumpStatus: String
    let axillaryFoldDiff: CGFloat
    let axillaryStatus: String
    let trunkRotationAngle: CGFloat
    let rotationStatus: String
    let trunkOffset: CGFloat
    let scapularWingingDiff: CGFloat
    let pelvicObliquity: CGFloat
    let rotationRiskScore: CGFloat
}

func analyzeRotation(
    landmarks: [[String: CGFloat]],
    imageSize: CGSize,
    ageGroup: String = "under15"
) -> RotationResult? {
    guard landmarks.count >= 27 else { return nil }

    let detector = PoseDetector.shared

    let lSh = detector.landmark(landmarks, at: 11) ?? .zero
    let rSh = detector.landmark(landmarks, at: 12) ?? .zero
    let lHip = detector.landmark(landmarks, at: 23) ?? .zero
    let rHip = detector.landmark(landmarks, at: 24) ?? .zero
    let lElbow = detector.landmark(landmarks, at: 13) ?? .zero
    let rElbow = detector.landmark(landmarks, at: 14) ?? .zero

    let midHipX = (lHip.x + rHip.x) / 2

    // Rib hump: lateral protrusion from hip midline
    let lLateral = abs(lSh.x - midHipX)
    let rLateral = abs(rSh.x - midHipX)
    let ribHumpProxy = abs(lLateral - rLateral)

    // Axillary fold difference
    let axillaryDiff = abs(lElbow.y - rElbow.y)

    // Trunk rotation angle: shoulder line vs hip line
    let shAngle = atan2(rSh.y - lSh.y, rSh.x - lSh.x) * 180 / .pi
    let hipAngle = atan2(rHip.y - lHip.y, rHip.x - lHip.x) * 180 / .pi
    var rotationAngle = abs(shAngle - hipAngle)
    if rotationAngle > 180 { rotationAngle = 360 - rotationAngle }

    // Trunk offset
    let midShX = (lSh.x + rSh.x) / 2
    let trunkOffset = midShX - midHipX

    // Scapular winging proxy
    let lScap = abs(lElbow.x - lSh.x)
    let rScap = abs(rElbow.x - rSh.x)
    let scapDiff = abs(lScap - rScap)

    // Pelvic obliquity
    let pelvicObliquity = abs(lHip.y - rHip.y)

    let thresholds = rotationAgeThresholds[ageGroup] ?? .default

    let ribHumpStatus = ribHumpProxy < thresholds.ribHumpPx ? "good" : "needs_improvement"
    let rotationStatus = rotationAngle < thresholds.rotationAngleDeg ? "good" : "needs_improvement"
    let axillaryStatus = axillaryDiff < thresholds.axillaryDiffPx ? "good" : "needs_improvement"

    // Risk score
    var riskScore: CGFloat = 0
    if ribHumpProxy > thresholds.ribHumpPx {
        riskScore += min(40, (ribHumpProxy / thresholds.ribHumpPx) * 20)
    }
    if rotationAngle > thresholds.rotationAngleDeg {
        riskScore += min(30, (rotationAngle / thresholds.rotationAngleDeg) * 15)
    }
    if axillaryDiff > thresholds.axillaryDiffPx {
        riskScore += min(20, (axillaryDiff / thresholds.axillaryDiffPx) * 10)
    }
    if scapDiff > thresholds.ribHumpPx {
        riskScore += min(10, (scapDiff / thresholds.ribHumpPx) * 5)
    }

    return RotationResult(
        ribHumpProxy: ribHumpProxy.roundedTo(2),
        ribHumpStatus: ribHumpStatus,
        axillaryFoldDiff: axillaryDiff.roundedTo(2),
        axillaryStatus: axillaryStatus,
        trunkRotationAngle: rotationAngle.roundedTo(2),
        rotationStatus: rotationStatus,
        trunkOffset: trunkOffset.roundedTo(2),
        scapularWingingDiff: scapDiff.roundedTo(2),
        pelvicObliquity: pelvicObliquity.roundedTo(2),
        rotationRiskScore: riskScore.roundedTo(1)
    )
}
